src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -707,196 +705,90 @@ static VaList empty() {
-            public boolean isPointer() {
-                return this == POINTER;
-            }
-
-            private static DynamicConstantDesc<ValueLayout> findBSM(String fieldName) {
-                return DynamicConstantDesc.ofNamed(
-                    BSM_GET_STATIC_FINAL,
-                    fieldName,
-                    CValueLayout.class.describeConstable().orElseThrow(),
-                    CLinker.class.describeConstable().orElseThrow()
-                );
-            }
-        }
-
-        private final Kind kind;
-
+        CHAR(true),
         /**
-         * CValueLayout constructor
-         *
-         * @param kind the kind of CValueLayout
-         * @param order the byte order of the layout
-         * @param bitSize the size, in bits, of the layout
-         * @param bitAlignment the alignment, in bits, of the layout
-         * @param attributes the attribute map of this layout
+         * A kind corresponding to the C {@code short} type
          */
-        public CValueLayout(Kind kind, ByteOrder order, long bitSize, long bitAlignment,
-                            Map<String, Constable> attributes) {
-            super(order, bitSize, bitAlignment, attributes);
-            this.kind = kind;
-        }
-
+        SHORT(true),
         /**
-         * Accessor for the kind of this layout
-         *
-         * @return the kind
+         * A kind corresponding to the C {@code int} type
          */
-        public final Kind kind() {
-            return kind;
-        }
-
+        INT(true),
         /**
-         * {@inheritDoc}
+         * A kind corresponding to the C {@code long} type
          */
-        @Override
-        public CValueLayout withOrder(ByteOrder order) {
-            return new CValueLayout(kind, order, bitSize(), alignment, attributes);
-        }
-
+        LONG(true),
         /**
-         * {@inheritDoc}
+         * A kind corresponding to the C {@code long long} type
          */
-        @Override
-        public CValueLayout withName(String name) {
-            return (CValueLayout) super.withName(name);
-        }
-
+        LONGLONG(true),
         /**
-         * {@inheritDoc}
+         * A kind corresponding to the C {@code float} type
          */
-        @Override
-        public CValueLayout withBitAlignment(long alignmentBits) {
-            return (CValueLayout) super.withBitAlignment(alignmentBits);
-        }
-
+        FLOAT(false),
         /**
-         * {@inheritDoc}
+         * A kind corresponding to the C {@code double} type
          */
-        @Override
-        public CValueLayout withAttribute(String name, Constable value) {
-            return (CValueLayout) super.withAttribute(name, value);
-        }
+        DOUBLE(false),
+        /**
+         * A kind corresponding to the C {@code long double} type
+         */
+        LONGDOUBLE(false),
+        /**
+         * A kind corresponding to the a C pointer type
+         */
+        POINTER(false);
```

Comment:
Pointer or pointer-sized? We'd also use this one for `size_t` for instance, right?

Reply:
This is pointer; e.g. values of this layout will be associated with the MemoryAddress.class carrier. So, no, for size_t, you need something different (either LONG or LONGLONG, depending on the platform). We could probably add a SIZE_T in there, since it's frequent enough.